flake.nix

```diff
@@ -2,64 +2,51 @@
-    nixpkgs-unstable.url = "github:nixos/nixpkgs/nixpkgs-unstable";
     nixpkgs.url = "github:nixos/nixpkgs/nixos-23.05";
+    nixpkgs-unstable.url = "github:nixos/nixpkgs/nixpkgs-unstable";
+
     systems.url = "github:nix-systems/default";
+    flake-parts.url = "github:hercules-ci/flake-parts";
   };
 
-  outputs = { self, flake-utils, nixpkgs-unstable, nixpkgs, ... }:
-    flake-utils.lib.eachDefaultSystem
-      (system:
-        let
-          pkgs = nixpkgs.legacyPackages.${system};
-          pkgs-unstable = nixpkgs-unstable.legacyPackages.${system};
-          versions = builtins.fromJSON (builtins.readFile ./versions.json);
-        in
-        {
-          devShells.default = pkgs.mkShell {
-            buildInputs = [
-              pkgs-unstable.black
-              (pkgs-unstable.python3.withPackages
-                (ps: [
-                  ps.pygithub
-                  ps.semver
-                ])
-              )
-              pkgs-unstable.nix-prefetch
-              pkgs.nodejs
-              pkgs.rubyPackages.dotenv
-            ];
-          };
-          # https://github.com/NixOS/nix/issues/7165
-          checks = self.packages.${system};
-          packages = builtins.mapAttrs
-            (version: { hash, vendorHash }: self.lib.buildTerraform {
-              inherit system version hash vendorHash;
-            })
-            versions;
-        }) // {
-      lib.buildTerraform = { system, version, hash, vendorHash }:
-        let
-          pkgs = nixpkgs.legacyPackages.${system};
-          pkgs-unstable = nixpkgs-unstable.legacyPackages.${system};
-        in
-        # https://www.hashicorp.com/blog/hashicorp-adopts-business-source-license
-        if builtins.compareVersions version "1.6.0" >= 0
-        then
-        # https://github.com/NixOS/nixpkgs/blob/nixpkgs-unstable/pkgs/applications/networking/cluster/terraform/default.nix
-          pkgs-unstable.mkTerraform
-            {
-              inherit version hash vendorHash;
-              patches = [ "${nixpkgs-unstable}/pkgs/applications/networking/cluster/terraform/provider-path-0_15.patch" ];
-            }
-        else
-        # https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/applications/networking/cluster/terraform/default.nix
-          pkgs.mkTerraform {
-            inherit version hash vendorHash;
-            patches = [ "${nixpkgs}/pkgs/applications/networking/cluster/terraform/provider-path-0_15.patch" ];
-          };
+  outputs = { self, flake-parts, ... }@inputs: flake-parts.lib.mkFlake { inherit inputs; } {
+    imports = [
+      inputs.flake-parts.flakeModules.easyOverlay
+    ];
+    systems = import inputs.systems;
+
+    perSystem = { config, pkgs, pkgs-unstable, system, ... }: {
+      _module.args = {
+        pkgs = import inputs.nixpkgs {
+          inherit system;
+        };
+        pkgs-unstable = import inputs.nixpkgs-unstable {
+          inherit system;
+          config.allowUnfree = true;
```

Comment:
I am not sure about this. @sestrella please advise.
On the one hand, enabling this by default makes `overlay` and `packages` work for all versions out of the box. Also if someone were to use `overlay` with, let's say Home-Manager instead of a shell, they would need to run `home-manager switch` with `NIXPKGS_ALLOW_UNFREE` or use `lib.buildTerraform` to provide modifier `nixpkgs`.

On the other - we're deciding on behalf of a consumer.

Reply:
I had another idea: we could make two subsets of `packages`: `free` and `unfree` or something like that. This way we could have both the convenience of things working out of the box and also a need of conscious choice from the user to use an unfree license.

Reply:
@pedorich-n regarding:

> On the other - we're deciding on behalf of a consumer.

I agree that it is preferable for users to explicitly agree to use unfree packages on their project; however, the error message produced by Nix when someone tries to use an unfree package is quite descriptive, so I don't believe there is much that needs to be done in that regard to help users fix an issue if they decide to use unfree packages. In that regard, I believe I am more included in leaving the structure of packages flat for now. Please let me know your thoughts.

Reply:
The only thing that bothers me is that there's no easy way to allow the unfree from within the nix. You'd have to either pass an env variable or build Terraform yourself using `mkTerraform` directly. 
But all the solutions I can think of aren't nice. 😕

Reply:
> there's no easy way to allow the unfree from within the nix

Although I agree with that statement, unfortunately, I believe that has always been the case for unfree packages. Therefore, I believe that Nix users shouldn't be surprised by the removal of `config.allowUnfree = true`, and even for those who are unfamiliar with Nix, the error message it generates is pretty instructive. That being said, I'll propose the following change:

```nix
_module.args = {
  pkgs-unstable = inputs.nixpkgs-unstable.legacyPackages.${system}
}
```

Reply:
Done